rfcs/0000-socket-proxy-kinesis-json.md

```diff
@@ -0,0 +1,168 @@
+- Feature Name: socket-proxy-kinesis-json
+- Start Date: 2021-06-02
+- RFC PR: mbta/technology-docs#0000
+- Asana task: [Secure connection between socket_proxy and rtr](https://app.asana.com/0/881264583703207/1200290609745412/f)
+- Status: Proposed
+
+# Summary
+
+Send OCS messages as JSON-formatted CloudEvents to AWS Kinesis
+
+# Motivation
+
+Currently, SocketProxy proxies OCS message to each RTR instance running in AWS. This method of sending Occupancy Control System (OCS) messages has several issues we are trying to address:
+- insecure: the messages travel unencrypted over the public internet
+- insecure: anyone from the MBTA network can forge OCS messages
+- unreliable: if an RTR instance isn't running (crashed, or during a restart/redeploy) when a message comes in, that message is lost
+- unreliable: if the production RTR instance isn't running, then the log of messages written to S3 (and used by OPMI) will not have those messages
+- missing metadata: the messages do not include the date they were sent, or the time the messages were received by SocketProxy 
+
+To address these issues, we'll update SocketProxy to write the OCS messages (along with additional metadata) as JSON CloudEvent messages to an AWS Kinesis stream. The partition key will be based on the type of message and the line.
```

Comment:
The reason I kept this in SocketProxy (even if the name is bad for a time) is so that we can run both the Kinesis-based RTR instances and the current TCP instances in parallel during development. wdyt?

The partition key is required, but we could set it to a random string each time if we didn't care about making sure an ordering is preserved. Or we could use something like `host:port` of the connecting socket, so all the messages from a given connection are routed to the same shard?

Reply:
> The reason I kept this in SocketProxy (even if the name is bad for a time) is so that we can run both the Kinesis-based RTR instances and the current TCP instances in parallel during development. wdyt?

This will mean repeated deploys of in-development branches to effectively "prod" Socket Proxy. We only have one real environment because the TCP sockets on the OCC end are configured to send to just a single IP/port.

However, one development flow we've used before with success is to have *another* separate app running on some port on opstech3, and having Socket Proxy proxy to `127.0.0.1:[that port]`.

> The partition key is required, but we could set it to a random string each time if we didn't care about making sure an ordering is preserved. Or we could use something like host:port of the connecting socket, so all the messages from a given connection are routed to the same shard?

Yep, I like the idea of `host:port` like in the RFC now. I'm not sure of the semantics of required ordering in RTR, but what we've effectively been using thus far is ordered-per-socket, with sockets interleaved. So `host:port` should maintain that.

Do listeners need to specify the shard? I think the host and/or port can change over time. And would the host and port be considered sensitive information?

Reply:
- good point on the redeploying SocketProxy: I'll update to indicate that this will be a new application, and we'll update the SocketProxy configuration to point at it.
- Listeners do need to specify the shard but that's true regardless of the partition key. If we're concerned about the host being part of the data, then we can hash it on our end, and use the hashed value as the partition key. That maintains the same ordering, without exposing the host in the data.